editoast/src/views/infra/mod.rs

```diff
@@ -158,6 +173,60 @@ async fn create(db_pool: Data<DbPool>, data: Json<InfraName>) -> Result<impl Res
     Ok(HttpResponse::Created().json(infra))
 }
 
+/// Duplicate an infra
+#[post("/clone")]
+async fn clone(
+    infra: Path<i64>,
+    db_pool: Data<DbPool>,
+    new_name: Query<InfraName>,
+) -> Result<Json<i64>> {
+    let db_pool_ref = db_pool.get_ref();
+    let db_pool_clone = db_pool_ref.clone();
+    let mut future: Vec<PinResult> = vec![];
+
+    let infra = infra.into_inner();
+    let cloned_infra = block::<_, Result<_>>(move || {
+        let name = new_name.name.clone();
+        let mut conn = db_pool_clone.get().expect("Failed to get DB connection");
+        Infra::clone(infra, &mut conn, name)
+    })
+    .await
+    .unwrap()?;
+
+    for object in ObjectType::iter() {
+        let model_table = object.get_table();
+        let db_pool_clone = db_pool_ref.clone();
+        let model = block::<_, Result<_>>(move || {
+            let mut conn = db_pool_clone.get().expect("Failed to get DB connection");
+            Ok(sql_query(format!(
+                "INSERT INTO {model_table}(id, obj_id,data,infra_id) SELECT nextval('{model_table}_id_seq'), obj_id,data,$1 FROM {model_table} WHERE infra_id=$2"
+            ))
+            .bind::<BigInt, _>(cloned_infra.id)
+            .bind::<BigInt, _>(infra)
+            .execute(&mut conn))
+        });
+        future.push(Box::pin(model));
+
+        if object != ObjectType::SwitchType && object != ObjectType::Route {
+            let layer_table = object.get_geometry_layer_table().unwrap().to_string();
+            let db_pool_clone = db_pool_ref.clone();
+            let layer = block::<_, Result<_>>(move || {
+                let mut conn = db_pool_clone.get().expect("Failed to get DB connection");
+                Ok(sql_query(format!(
+                    "INSERT INTO {layer_table}(id, obj_id,geographic,schematic,infra_id) SELECT nextval('{layer_table}_id_seq'), obj_id,geographic,schematic,$1 FROM {layer_table} WHERE infra_id=$2"
+                ))
+                .bind::<BigInt, _>(cloned_infra.id)
+                .bind::<BigInt, _>(infra)
+                .execute(&mut conn))
+            });
+            future.push(Box::pin(layer));
+        }
+    }
+    join_all(future).await;
```

Comment:
You need to check that the futures didn't failed.

```suggestion
    // Unwrap blocking error
    let res: Vec<_> = join_all(futures).await.collect().unwrap();
    // Check for diesel errors
    res.collect::<Result<_>>()?;
```